File: dfs/algo.py
def maxDepth(self, root):
    if not root:
        return 0
    return 1 + max(self.maxDepth(root.left), self.maxDepth(root.right))

def minDepth(self, root):
    if not root:
        return 0
    left = self.minDepth(root.left)
    right = self.minDepth(root.right)
    return left + right + 1 if (left == 0 or right == 0) else min(left, right) + 1

def generateParenthesis(self, n):
    self.list = []
    self._gen(0, 0, n, "")
    return self.list

    def _gen(self, left, right, n, result):
        if left == n and right == n:
            self.list.append(result)
            return
        if left < n:
            self._gen(left + 1, right, n, result + '(')
        if left > right and right < n:
            self._gen(left, right + 1, n, result + ')')

File: dfs/algo.java
public int maxDepth(TreeNode root) {
    return null == root ? 0 : 1 + Math.max(maxDepth(root.left), maxDepth(root.right));
}

public int minDepth(TreeNode root) {
    if (root == null) return 0;
    int left = minDepth(root.left);
    int right = minDepth(root.right);
    return (left == 0 || right == 0) ? left + right + 1 : Math.min(left, right) + 1;
}

public List<String> generateParenthesis(int n) {
    List<String> result = new ArrayList<>();
    gen(0, 0, n, "", result);
    return result;
}

private void gen(int left, int right, int n, String str, List<String> result) {
    if (left == n && right == n) {
        result.add(str);
        return;
    }
    if (left < n) {
        gen(left+1,right,n,str + "(",result);
    }
    if (left > right && right < n) {
        gen(left,right+1,n,str + ")",result);
    }
}